sweep/txgenerator.go

```diff
@@ -123,6 +134,87 @@ func generateInputPartitionings(sweepableInputs []txInput,
 	return sets, nil
 }
 
+// addWalletInputs retrieves utxos from the wallet and tries adding as many as
+// required to bring the tx output value above the given minimum.
+func addWalletInputs(inputSet *txInputSet, minOutputAmt btcutil.Amount,
+	wallet Wallet) (bool, error) {
+
+	// Retrieve wallet utxos. Only consider confirmed utxos to prevent
+	// problems around RBF rules for unconfirmed inputs.
+	utxos, err := wallet.ListUnspentWitness(1, math.MaxInt32)
+	if err != nil {
+		return false, err
+	}
+
+	for _, utxo := range utxos {
+		input, err := createWalletTxInput(utxo)
+		if err != nil {
+			return false, err
+		}
+
+		// If the wallet input isn't positively-yielding at this fee
+		// rate, skip it.
+		if !inputSet.add(input) {
+			continue
+		}
+
+		// Return if we've reached the minimum output amount.
+		if inputSet.outputValue >= minOutputAmt {
+			return true, nil
+		}
+	}
+
+	// We were not able to reach the minimum output amount.
+	return false, nil
+}
+
+// createWalletTxInput converts a wallet utxo into an object that can be added
+// to the other inputs to sweep.
+func createWalletTxInput(utxo *lnwallet.Utxo) (txInput, error) {
+	var witnessType input.WitnessType
+	switch utxo.AddressType {
+	case lnwallet.WitnessPubKey:
+		witnessType = input.WitnessKeyHash
+	case lnwallet.NestedWitnessPubKey:
+		witnessType = input.NestedWitnessKeyHash
+	default:
+		return nil, fmt.Errorf("unknown input witness")
+	}
+
+	signDesc := &input.SignDescriptor{
+		Output: &wire.TxOut{
+			PkScript: utxo.PkScript,
+			Value:    int64(utxo.Value),
+		},
+		HashType: txscript.SigHashAll,
+	}
+
+	// A height hint doesn't need to be set, because we don't monitor these
+	// inputs for spend.
+	heightHint := uint32(0)
+
+	walletInput := input.NewBaseInput(
+		&utxo.OutPoint, witnessType, signDesc, heightHint,
+	)
+
+	return &walletTxInput{
+		Input: walletInput,
+	}, nil
+}
+
+// walletTxInput wraps a wallet input in the same way other sweeper inputs are
+// wrapped.
+type walletTxInput struct {
```

Comment:
instead of having  a special `walletTxInput` type, what do you think about having the `txInput` define a `gain` or similar. It would be `0` for wallet inputs, and the actual  output value for other types we sweep.

When adding the wallet input, it would check that the total `gain-fee` is positive.

Reply:
I still need to create an object that conforms to the `txInput` interface. I don't want to use `pendingInput` for this.

Reply:
Gain could also work, but it would then always be either zero or the input value. A bool would do the same thing. But adding a bool also seemed redundant if I can just type check here.

Reply:
Decided to make a `fromWallet` boolean for now.